### PR TITLE
Best practices for contributing to tskit

### PR DESCRIPTION
**Pull request by awohns**
I thought I would document some suggestions for best practices in contributing to tskit-dev. These are based on my experience with kc_distance.

### CODE REVIEW

File: docs/development.rst [hunk 1]
```diff
@@ -83,3 +84,59 @@ install the bleeding-edge package directly from the tskit github repository requ
 to specify ``subdirectory=python``:
 
 ``python3 -m pip install git+https://github.com/tskit-dev/tskit.git#subdirectory=python``
+
+***********************************************
+Best Practices for adding to the tskit codebase
```
Comment (jeromekelleher) [inline docs/development.rst]: This is a bit long for a title (these get rendered in the left navbar). How about, "Development workflow" or "Feature lifecycle" or something?

File: docs/development.rst [hunk 2]
```diff
@@ -83,3 +84,59 @@ install the bleeding-edge package directly from the tskit github repository requ
...
+***********************************************
+
+The following is a rough guide of best practices for contributing a function to the
+tskit codebase.
+We follow the example of the `kc_distance function` 
+
+0.  Open an `issue <http://github.com/tskit-dev/tskit/issues>`_ with your proposed
+    functionality. If consensus is reached that your proposed addition shoudl be
```
Comment (jeromekelleher) [inline docs/development.rst]: Typo, shoudl

File: docs/development.rst [hunk 3]
```diff
@@ -83,3 +84,59 @@ install the bleeding-edge package directly from the tskit github repository requ
...
+    `c directory <https://github.com/tskit-dev/tskit/tree/master/c/tskit>`_.
+    Your function will probably go in 
+    `trees.c <https://github.com/tskit-dev/tskit/blob/master/c/tskit/trees.c>`_.
+5.  Write a few tests for your function in C: again, write your tests in  
+    `tskit/c/tests/test_tree.c <https://github.com/tskit-dev/tskit/blob/master/c/tests/test_trees.c>`_.
+    The key here is code coverage, you don't need to worry as much about covering every
+    corner case, as we will proceed to link this function to the Python tests you
+    wrote earlier.
+6.  Create a low-level definition of your function using Python's C API: This will
+    go in `tskit/python/_tskit_module.c
```
Comment (jeromekelleher) [inline docs/development.rst]: _tskitmodule.c

File: docs/development.rst [hunk 4]
```diff
@@ -83,3 +84,59 @@ install the bleeding-edge package directly from the tskit github repository requ
...
+5.  Write a few tests for your function in C: again, write your tests in  
+    `tskit/c/tests/test_tree.c <https://github.com/tskit-dev/tskit/blob/master/c/tests/test_trees.c>`_.
+    The key here is code coverage, you don't need to worry as much about covering every
+    corner case, as we will proceed to link this function to the Python tests you
+    wrote earlier.
+6.  Create a low-level definition of your function using Python's C API: This will
+    go in `tskit/python/_tskit_module.c 
+    <https://github.com/tskit-dev/tskit/blob/master/python/_tskitmodule.c>`_.
+7.  Test your low-level implementation in `tskit/python/tests/test_lowlevel.py
+    <https://github.com/tskit-dev/tskit/blob/master/python/tests/test_lowlevel.py>`_.
```
Comment (jeromekelleher) [inline docs/development.rst]: , concentrating particularly on the *interface* (does the function behave correctly on malformed inputs), etc.

File: docs/development.rst [hunk 5]
```diff
@@ -83,3 +84,59 @@ install the bleeding-edge package directly from the tskit github repository requ
...
+    the kc_distance function lives in 
+    `tskit/python/tskit/trees.py, for example
+    <https://github.com/tskit-dev/tskit/blob/master/python/tskit/trees.py>`_.
+9.  Modify your Python tests to test the new C-linked function. If you followed
+    the example of other tests, you might need to only add a single line of code
+    here.
+10. Write a docstring for your function in the Python API: for instance, the kc_distance
+    docstring is in 
+    `tskit/python/tskit/trees.py 
+    <https://github.com/tskit-dev/tskit/blob/master/python/tskit/trees.py>`_.
```
Comment (jeromekelleher) [inline docs/development.rst]: Make sure that the docstring renders correctly by building the documentation (do we have a section for the documentation here? We could make a stub and link to it, and make an issue on github for filling out this stub)

File: docs/development.rst (hunk 1 above)
Comment (andrewkern) [reply docs/development.rst]: or "Best Practices for Development"

**Comment by jeromekelleher** [comment]
How's this looking @awohns?

File: docs/development.rst (hunk 5 above)
Comment (awohns) [reply docs/development.rst]: Good idea!

**Comment by jeromekelleher** [comment]
Sphinxarg might have been added as a dev requirement since then.``pip install -r requirements/developments.txt`` should fix it.

**Comment by awohns** [comment]
Pushed changes addressing issues opened by @jeromekelleher above. Also created issue #454 to document the "how to add to the documentation" stub.

**Comment by jeromekelleher** [comment]
LGTM, thanks @awohns! We're ready to merge after a quick rebase, I think.

**Comment by awohns** [comment]
Disregard the above, I merged after the successful rebase and messed up the branch, fixing now

**Comment by awohns** [comment]
All set!

**Comment by jeromekelleher** [comment]
Done, thanks @awohns!
